services/github/src/queue-lock.js

```diff
@@ -0,0 +1,45 @@
+/**
+ * Implements locked queue to allow one routine running at a time
+ */
+class QueuedLock {
+  constructor() {
+    this.locks = new Map();
+    this.queue = {};
+  }
+
+  acquire(name) {
+    if (!this.queue[name]) {
+      this.queue[name] = [];
+    }
+
+    const [ promise, resolver ] = this._createPromise();
+
+    if (!this.locks.has(name)) {
+      this.locks.set(name, true);
+      resolver();
+    } else {
+      this.queue[name].push(resolver);
+    }
+
+    return promise;
+  }
+
+  release(name) {
+    const nextResolver = this.queue[name].shift();
```

Comment:
This is the only part that can cause issues, if client code doesn't call `.release()` all pending callers will be stuck. 
At the moment status handler ensures that it always calls it before the end.

For the future we might want to implement some fail-safe watchdog timers to force-release lock after some timeout

Reply:
I don't know if a timer is the right fix here.  If something is stuck and taking forever (like the GitHub API), probably waiting for it to fail is the right approach.  The risk is that a `return` gets added to the function without a corresponding `release`.  One way to avoid that is to have the `acquire` take an async callback which is called with the lock held, and unconditionally release the lock when the callback returns.

Reply:
Yeah, I was also thinking about this callback approach, but didn't want to make code more complicated on the handler side